slither/detectors/variables/uninitialized_state_variables.py

```diff
@@ -24,29 +28,36 @@ class UninitializedStateVarsDetection(AbstractDetector):
     IMPACT = DetectorClassification.HIGH
     CONFIDENCE = DetectorClassification.HIGH
 
-    def detect_uninitialized(self, contract):
-        # get all the state variables read by all functions
-        var_read = [f.state_variables_read for f in contract.all_functions_called + contract.modifiers]
-        # flat list
-        var_read = [item for sublist in var_read for item in sublist]
-        # remove state variable that are initiliazed at contract construction
-        var_read = [v for v in var_read if v.uninitialized]
-
-        # get all the state variables written by the functions
-        var_written = [f.state_variables_written for f in contract.all_functions_called + contract.modifiers]
-        # flat list
-        var_written = [item for sublist in var_written for item in sublist]
-
-        all_push = [f.apply_visitor(FindPush) for f in contract.functions]
-        # flat list
-        all_push = [item for sublist in all_push for item in sublist]
+    @staticmethod
+    def written_variables(contract):
+        ret = []
+        for f in contract.all_functions_called + contract.modifiers:
+            for n in f.nodes:
+                for ir in n.irs:
+                    if isinstance(ir, (Index, Member)):
+                        continue  # Don't consider Member and Index operations -> ReferenceVariable
+                    elif isinstance(ir, OperationWithLValue) and isinstance(ir.lvalue, StateVariable):
+                        ret.append(ir.lvalue)
+                    elif isinstance(ir, Assignment) and isinstance(ir.lvalue, ReferenceVariable):
```

Comment:
So the way the references are tracked will not catch all the cases (Reference can be the result of any operation, not only `Assignment`).

I simplified the API: https://github.com/trailofbits/slither/pull/58. From the `dev-refactor-nodes-info` branch,  you can directly uses  `node.state_variables_written` (it recovers the info from the IR directly). It should simplify your function

Though `node.state_variables_written` does not consider the write through a library, so you still need to iterate over the IR and check for `storage` variable, which your function already do.

So you can remove the three first `if` branch here, replace by `node.state_variables_written`, and only keep if branch of the line 46

Reply:
@montyly this definitely is a good improvement. I will wait till your PR lands in dev and will afterwards rebase mine with the changes needed.